File: BackEnd/Praza-Shop/src/main/java/org/example/prazashop/service/PedidoService.java
package org.example.prazashop.service;

import org.example.prazashop.model.dto.PedidoDto;
import org.example.prazashop.model.dto.PedidoConDetallesDto;

import java.util.List;

/**
 * Servicios de negocio relacionados con pedidos.
 */
public interface PedidoService {
    /**
     * Find all list.
     *
     * @return the list
     */
    List<PedidoDto> findAll();

    /**
     * Busca un pedido por su id. Lanza NoContentException si no se encuentra.
     *
     * @param id the id
     * @return pedido encontrado
     * @throws NoContentException si no se encuentra el pedido
     */
    PedidoDto findById(Long id);

    /**
     * Create pedido dto.
     *
     * @param pedido the pedido
     * @return the pedido dto
     */
    PedidoDto create(PedidoDto pedido);

    /**
     * Update pedido dto.
     *
     * @param id     the id
     * @param pedido the pedido
     * @return the pedido dto
     */
    PedidoDto update(Long id, PedidoDto pedido);

    /**
     * Delete by id.
     *
     * @param id the id
     */
    void deleteById(Long id);

    /**
     * Busca pedidos por id de cliente.
     *
     * @param clienteId id del cliente
     * @return lista de pedidos
     */
    List<PedidoDto> findByClienteId(Long clienteId);

    /**
     * Busca pedidos con detalles por id de cliente.
     *
     * @param clienteId id del cliente
     * @return lista de pedidos con sus detalles
     */
    List<PedidoConDetallesDto> findByClienteIdConDetalles(Long clienteId);

    /**
     * Busca pedidos con detalles por id de negocio.
     *
     * @param negocioId id del negocio
     * @return lista de pedidos con sus detalles
     */
    List<PedidoConDetallesDto> findByNegocioIdConDetalles(Long negocioId);

    /**
     * Busca pedidos por id de negocio.
     *
     * @param negocioId id del negocio
     * @return lista de pedidos
     */
    List<PedidoDto> findByNegocioId(Long negocioId);

    /**
     * Crea un pedido completo con detalles en una transacción atómica.
     *
     * @param request request con clienteId, negocioId y detalles
     * @return el pedido creado con detalles
     */
    PedidoConDetallesDto crearPedidoCompleto(org.example.prazashop.model.dto.PedidoCreacionRequest request);

    /**
     * Busca pedidos con filtros y paginación.
     *
     * @param estado estado del pedido (opcional)
     * @param fechaDesde fecha mínima (opcional)
     * @param fechaHasta fecha máxima (opcional)
     * @param precioDesde precio mínimo (opcional)
     * @param precioHasta precio máximo (opcional)
     * @param ordenar criterio de ordenamiento: fecha_asc, fecha_desc, total_asc, total_desc
     * @param pagina número de página (0-indexed)
     * @param tamaño cantidad de resultados por página
     * @return response con paginación
     */
    org.example.prazashop.model.dto.PedidoSearchResponse buscarPedidos(
            String estado,
            java.time.LocalDateTime fechaDesde,
            java.time.LocalDateTime fechaHasta,
            Double precioDesde,
            Double precioHasta,
            String ordenar,
            Integer pagina,
            Integer tamaño
    );

    /**
     * Obtiene la entidad Pedido por ID (sin convertir a DTO).
     *
     * @param id id del pedido
     * @return entidad Pedido
     */
    org.example.prazashop.model.entity.Pedido findEntityById(Long id);

    /**
     * Convierte una entidad Pedido a PedidoConDetallesDto.
     *
     * @param pedido entidad a convertir
     * @return PedidoConDetallesDto
     */
    org.example.prazashop.model.dto.PedidoConDetallesDto toDtoConDetalles(org.example.prazashop.model.entity.Pedido pedido);
}

File: FrontEnd/praza_shop/lib/services/pedido_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'api_service.dart';
import '../models/pedido_dto.dart';
import '../models/pedido_con_detalles_dto.dart';

class PedidoService {
  final ApiService api;
  PedidoService(this.api);

  Future<List<PedidoDto>> getAll() async {
    final uri = Uri.parse('${api.baseUrl}/api/pedidos');
    final res = await http.get(uri, headers: api.headers(jsonBody: false));
    if (res.statusCode == 200) {
      final data = json.decode(res.body) as List<dynamic>;
      return data.map((e) => PedidoDto.fromJson(e as Map<String, dynamic>)).toList();
    }
    throw Exception('Get pedidos failed: ${res.statusCode} ${res.body}');
  }

  Future<PedidoDto> getById(int id) async {
    final uri = Uri.parse('${api.baseUrl}/api/pedidos/$id');
    final res = await http.get(uri, headers: api.headers(jsonBody: false));
    if (res.statusCode == 200) return PedidoDto.fromJson(json.decode(res.body));
    throw Exception('Get pedido failed: ${res.statusCode} ${res.body}');
  }

  Future<PedidoDto> create(PedidoDto p) async {
    final uri = Uri.parse('${api.baseUrl}/api/pedidos');
    final res = await http.post(uri, headers: api.headers(), body: json.encode(p.toJson()));
    if (res.statusCode == 200 || res.statusCode == 201) return PedidoDto.fromJson(json.decode(res.body));
    throw Exception('Create pedido failed: ${res.statusCode} ${res.body}');
  }

  Future<PedidoDto> update(int id, PedidoDto p) async {
    final uri = Uri.parse('${api.baseUrl}/api/pedidos/$id');
    final res = await http.put(uri, headers: api.headers(), body: json.encode(p.toJson()));
    if (res.statusCode == 200) return PedidoDto.fromJson(json.decode(res.body));
    throw Exception('Update pedido failed: ${res.statusCode} ${res.body}');
  }

  Future<void> delete(int id) async {
    final uri = Uri.parse('${api.baseUrl}/api/pedidos/$id');
    final res = await http.delete(uri, headers: api.headers(jsonBody: false));
    if (res.statusCode != 200 && res.statusCode != 204) throw Exception('Delete pedido failed: ${res.statusCode} ${res.body}');
  }

  Future<List<PedidoDto>> findByNegocioId(int negocioId) async {
    final uri = Uri.parse('${api.baseUrl}/api/pedidos/negocio/$negocioId');
    final res = await http.get(uri, headers: api.headers(jsonBody: false));
    if (res.statusCode == 200) {
      final data = json.decode(res.body) as List<dynamic>;
      return data.map((e) => PedidoDto.fromJson(e as Map<String, dynamic>)).toList();
    }
    throw Exception('Get pedidos by negocio failed: ${res.statusCode} ${res.body}');
  }

  Future<List<PedidoDto>> findByClienteId(int clienteId) async {
    final uri = Uri.parse('${api.baseUrl}/api/pedidos/cliente/$clienteId');
    final res = await http.get(uri, headers: api.headers(jsonBody: false));
    if (res.statusCode == 200) {
      final data = json.decode(res.body) as List<dynamic>;
      return data.map((e) => PedidoDto.fromJson(e as Map<String, dynamic>)).toList();
    }
    throw Exception('Get pedidos by cliente failed: ${res.statusCode} ${res.body}');
  }

  Future<List<PedidoConDetallesDto>> findByClienteIdConDetalles(int clienteId) async {
    final uri = Uri.parse('${api.baseUrl}/api/pedidos/cliente/$clienteId/detalles');
    final res = await http.get(uri, headers: api.headers(jsonBody: false));
    if (res.statusCode == 200) {
      final data = json.decode(res.body) as List<dynamic>;
      return data.map((e) => PedidoConDetallesDto.fromJson(e as Map<String, dynamic>)).toList();
    }
    throw Exception('Get pedidos with details by cliente failed: ${res.statusCode} ${res.body}');
  }

  /// Crea un pedido completo con detalles en una sola llamada atómica
  Future<PedidoConDetallesDto> crearPedidoCompleto({
    required int clienteId,
    required int negocioId,
    required List<Map<String, dynamic>> detalles, // [{productoId, cantidad}, ...]
  }) async {
    final uri = Uri.parse('${api.baseUrl}/api/pedidos/crear-completo');
    final body = json.encode({
      'clienteId': clienteId,
      'negocioId': negocioId,
      'detalles': detalles,
    });
    final res = await http.post(uri, headers: api.headers(), body: body);
    if (res.statusCode == 200 || res.statusCode == 201) {
      return PedidoConDetallesDto.fromJson(json.decode(res.body));
    }
    throw Exception('Crear pedido completo failed: ${res.statusCode} ${res.body}');
  }

  /// Busca pedidos con filtros y paginación
  Future<Map<String, dynamic>> buscarPedidos({
    String? estado,
    String? fechaDesde,
    String? fechaHasta,
    double? precioDesde,
    double? precioHasta,
    String ordenar = 'fecha_desc',
    int pagina = 0,
    int tamano = 10,
  }) async {
    final params = <String, String>{
      'ordenar': ordenar,
      'pagina': pagina.toString(),
      'tamano': tamano.toString(),
    };

    if (estado != null) params['estado'] = estado;
    if (fechaDesde != null) params['fechaDesde'] = fechaDesde;
    if (fechaHasta != null) params['fechaHasta'] = fechaHasta;
    if (precioDesde != null) params['precioDesde'] = precioDesde.toString();
    if (precioHasta != null) params['precioHasta'] = precioHasta.toString();

    final uri = Uri.parse('${api.baseUrl}/api/pedidos/buscar').replace(queryParameters: params);
    final res = await http.get(uri, headers: api.headers(jsonBody: false));
    if (res.statusCode == 200) {
      return json.decode(res.body) as Map<String, dynamic>;
    }
    throw Exception('Buscar pedidos failed: ${res.statusCode} ${res.body}');
  }

  /// ===== NUEVOS ENDPOINTS USER-SCOPED (Fase 3) =====

  /// Obtiene los pedidos del cliente autenticado con paginación
  Future<Map<String, dynamic>> misPedidos({
    int pagina = 0,
    int tamano = 10,
  }) async {
    final params = <String, String>{
      'pagina': pagina.toString(),
      'tamaño': tamano.toString(),
    };

    final uri = Uri.parse('${api.baseUrl}/api/mi-compra/pedidos').replace(queryParameters: params);
    final res = await http.get(uri, headers: api.headers(jsonBody: false));
    if (res.statusCode == 200) {
      return json.decode(res.body) as Map<String, dynamic>;
    }
    throw Exception('Get mis pedidos failed: ${res.statusCode} ${res.body}');
  }

  /// Obtiene las ventas del negocio autenticado con paginación
  Future<Map<String, dynamic>> misVentas({
    int pagina = 0,
    int tamano = 10,
  }) async {
    final params = <String, String>{
      'pagina': pagina.toString(),
      'tamaño': tamano.toString(),
    };

    final uri = Uri.parse('${api.baseUrl}/api/mi-negocio/ventas').replace(queryParameters: params);
    final res = await http.get(uri, headers: api.headers(jsonBody: false));
    if (res.statusCode == 200) {
      return json.decode(res.body) as Map<String, dynamic>;
    }
    throw Exception('Get mis ventas failed: ${res.statusCode} ${res.body}');
  }

  /// Crea un pedido como cliente (auto-resuelve clienteId)
  Future<PedidoConDetallesDto> crearPedidoComoCliente({
    required int negocioId,
    required List<Map<String, dynamic>> detalles, // [{productoId, cantidad}, ...]
  }) async {
    final uri = Uri.parse('${api.baseUrl}/api/mi-compra/pedidos');
    final body = json.encode({
      'negocioId': negocioId,
      'detalles': detalles,
    });
    final res = await http.post(uri, headers: api.headers(), body: body);
    if (res.statusCode == 200 || res.statusCode == 201) {
      return PedidoConDetallesDto.fromJson(json.decode(res.body));
    }
    throw Exception('Crear pedido como cliente failed: ${res.statusCode} ${res.body}');
  }

  /// Actualiza el estado de un pedido (solo propietario del negocio)
  Future<PedidoConDetallesDto> actualizarEstadoPedido({
    required int pedidoId,
    required String nuevoEstado,
  }) async {
    final uri = Uri.parse('${api.baseUrl}/api/mi-negocio/ventas/$pedidoId/estado');
    final bodyMap = {
      'nuevoEstado': nuevoEstado,
    };
    final body = json.encode(bodyMap);
    print('DEBUG: Actualizando estado - URI: $uri');
    print('DEBUG: Body enviado: $body');
    final res = await http.put(uri, headers: api.headers(), body: body);
    print('DEBUG: Response status: ${res.statusCode}');
    print('DEBUG: Response body: ${res.body}');
    if (res.statusCode == 200) {
      return PedidoConDetallesDto.fromJson(json.decode(res.body));
    }
    throw Exception('Actualizar estado pedido failed: ${res.statusCode} ${res.body}');
  }
}
